src/aof.zig

```diff
@@ -0,0 +1,656 @@
+const std = @import("std");
+const assert = std.debug.assert;
+const os = std.os;
+
+const constants = @import("constants.zig");
+const vsr = @import("vsr.zig");
+const tb = @import("tigerbeetle.zig");
+
+const stdx = @import("stdx.zig");
+const IO = @import("io.zig").IO;
+const MessagePool = @import("message_pool.zig").MessagePool;
+const Message = @import("message_pool.zig").MessagePool.Message;
+const MessageBus = @import("message_bus.zig").MessageBusClient;
+const Storage = @import("storage.zig").Storage;
+const StateMachine = @import("state_machine.zig").StateMachineType(Storage, constants.state_machine_config);
+
+const Header = vsr.Header;
+const Account = tb.Account;
+const Transfer = tb.Transfer;
+const Client = vsr.Client(StateMachine, MessageBus);
+const log = std.log.scoped(.aof);
+
+const MAGIC_NUMBER: u128 = 312960301372567410560647846651901451202;
+
+/// On disk format for AOF Metadata
+pub const AOFMetadata = extern struct {
+    comptime {
+        assert(@bitSizeOf(AOFMetadata) == @sizeOf(AOFMetadata) * 8);
+    }
+
+    vsr_header: Header,
+    primary: u64,
+    replica: u64,
+};
+
+/// On disk format for full AOF Entry
+pub const AOFEntry = extern struct {
+    comptime {
+        assert(@bitSizeOf(AOFEntry) == @sizeOf(AOFEntry) * 8);
+    }
+
+    /// In case of extreme corruption, start each entry with a fixed random integer,
+    /// to allow skipping over corrupted entries.
+    magic_number: u128 = MAGIC_NUMBER,
+
+    /// Arbitrary metadata we want to record
+    metadata: AOFMetadata,
+
+    /// The main Message to log. The actual length of the entire payload will be sector
+    /// aligned, so it might continue past body_size.
+    body_size: u64,
+    body: [constants.message_size_max]u8,
+
+    /// Reserved space at the end, in case we require padding past message_size_max.
+    reserved: [constants.sector_size]u8 = std.mem.zeroes([constants.sector_size]u8),
+
+    /// Calculate the actual length of the entry, from our fixed length struct. Make sure to
+    /// exclude the reserved padding at the end, as we use this in case of large messages.
+    pub fn calculate_padded_size(self: *AOFEntry) u64 {
+        const unpadded_size = @sizeOf(AOFEntry) - constants.message_size_max - constants.sector_size + self.body_size;
```

Comment:
The `AOFEntry` struct has a few other fields, ahead of what was `.body` - this calc is to work out the size of the entry that needs to be written to disk (since we don't need to write out all the 0s but we do need to be sector aligned).

I've changed it now to
`const truncated_size = @sizeOf(AOFEntry) - self.message.len + self.header().size;` and renamed the fn to `calculate_disk_size` if that's more clear?